File: src/core/Options/Handlers/Export.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using core.Shared;
using core.Shared.Adapters.CSV;

namespace core.Options
{
    public class Export
    {
        public class Query : RequestWithUserId<ExportResponse>
        {
            public Query(Guid userId) : base(userId)
            {
            }
        }

        public class Handler : HandlerWithStorage<Query, ExportResponse>
        {
            private ICSVWriter _csvWriter;

            public Handler(
                ICSVWriter csvWriter,
                IPortfolioStorage storage) : base(storage)
            {
                _csvWriter = csvWriter;
            }

            public override async Task<ExportResponse> Handle(Query request, CancellationToken cancellationToken)
            {
                var options = await _storage.GetOwnedOptions(request.UserId);

                var filename = CSVExport.GenerateFilename("options");

                return new ExportResponse(filename, CSVExport.Generate(_csvWriter, options));
            }
        }
    }
}

File: src/core.fs/Options/Export.fs
namespace core.fs.Options

open core.Shared.Adapters.CSV
open core.Shared.Adapters.Storage
open core.fs

module Export =

    type Query(userId:System.Guid) =
        member _.UserId = userId
        member _.Filename = "options"

    type Handler(storage:IPortfolioStorage, csvWriter:ICSVWriter) =
        interface IApplicationService

        member _.Handle(request:Query) =
            task {
                let! options = request.UserId |> storage.GetOwnedOptions;

                let csv = CSVExport.Generate(csvWriter, options);

                return ExportResponse(
                    request.Filename |> CSVExport.GenerateFilename,
                    csv
                );
            }